In [1]:
# Import  libraries

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

In [2]:
# Import dataset

titanic_data = pd.read_csv(r'C:\Users\rksan\Downloads\titanic_train.csv')
titanic_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
# Shape of the dataset

titanic_data.shape

(891, 12)

There are 891 Rows & 12 Columns, According to me columns=["PassangerId","Name","Ticket"] are not useful for the models. So, let's first drop it.

In [4]:
# Droping Unusable Columns

titanic_data.drop(columns = ["PassengerId", "Name", "Ticket"], axis=1, inplace=True)

print("Shape of the data --> ",titanic_data.shape)
titanic_data.head()

Shape of the data -->  (891, 9)


Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin Embarked
0         0       3    male  22.0      1      0   7.2500   NaN        S
1         1       1  female  38.0      1      0  71.2833   C85        C
2         1       3  female  26.0      0      0   7.9250   NaN        S
3         1       1  female  35.0      1      0  53.1000  C123        S
4         0       3    male  35.0      0      0   8.0500   NaN        S

Now our data is ready for EDA

In [5]:
# Quick information about dataset

titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
Survived    891 non-null int64
Pclass      891 non-null int64
Sex         891 non-null object
Age         714 non-null float64
SibSp       891 non-null int64
Parch       891 non-null int64
Fare        891 non-null float64
Cabin       204 non-null object
Embarked    889 non-null object
dtypes: float64(2), int64(4), object(3)
memory usage: 62.8+ KB


It seems that there are missing values in our dataset. Let's check how many missing values are available

In [6]:
# Missing Values

titanic_data.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Cabin       687
Embarked      2
dtype: int64

Columns = ["Age", "Cabin", "Embarked"] have null values. Let's check how many percentages (%) they are covering from the datset.

In [7]:
column_data = titanic_data.isna().sum().keys().tolist()
value_data = titanic_data.isna().sum().values.tolist()

missing_data = pd.DataFrame(list(zip(column_data, value_data)),columns=["Columns", "Missing Values"],index=list(range(1,10)))
missing_data["Missing %"] = round(((missing_data["Missing Values"]/titanic_data.shape[0])*100),2)

print("\nMissing Values for Titanic Data :-")
missing_data


Missing Values for Titanic Data :-


Columns  Missing Values  Missing %
1  Survived               0       0.00
2    Pclass               0       0.00
3       Sex               0       0.00
4       Age             177      19.87
5     SibSp               0       0.00
6     Parch               0       0.00
7      Fare               0       0.00
8     Cabin             687      77.10
9  Embarked               2       0.22

"Cabin" is covering 77.10 % of null values which is very high. and other are covering less null values.

In [8]:
# describing the dataset

titanic_data.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Survive is our target column.
some have age and some doesn't have age.
Age and Fare column have skewness. 
Survived is our target column and Pclass,SibSp & Parch looks as classification. 
So we need to focus on Age & Fare columns for filling null values.

In [9]:
 checking the "0" values if there are present

for col in titanic_data:
           print(col +"  "+str(titanic_data[titanic_data[col]==0].shape[0]))
        

SyntaxError: invalid syntax (<ipython-input-9-45abc234664f>, line 1)

I discussed about all column in which 0 is present but Fare column is not a classification column and it has 0 values. Let's analyze it.

In [ ]:
# checking Fare column where 0 values are present

titanic_data[titanic_data["Fare"]==0]

In [ ]:
# describing the categorical data

titanic_data.describe(include='object')

The frequency of Male is higher than Female, Males have covered 65% data. In Embarked S covered the 644 out of 889.

In [ ]:
# checking pairplot with Survived

plt.figure(figsize=(7,7))
sns.pairplot(titanic_data,hue="Survived")

In [ ]:
# Analyzing 

sns.catplot(x="Age",y="Sex",data=titanic_data,hue="Survived")

It seems that Females are more survive than Males

In [ ]:
## Handle the Missing Data


# Dropna Embarked
titanic_data.dropna(subset=["Embarked"],inplace=True)

# Filling Age with its mean value
titanic_data.fillna(titanic_data["Age"].mean(), inplace=True)

# Filling Cabin with 0
titanic_data.fillna(titanic_data["Cabin"].mode(),inplace=True)

In [ ]:
# Check Null Values again

titanic_data.isna().sum()

In [ ]:
#checking the skewness

titanic_data.skew()

As we discussed that SibSp & Parch are classifiation columns, So I have to go with Fare column for making skewness correct

In [ ]:
# separate the numerica columns and categorical columns

numerical = titanic_data.drop(columns=["Sex", "Cabin", "Embarked"])
categorical = titanic_data[["Sex", "Cabin", "Embarked"]]

In [ ]:
# visualizing the skewness with Distplot

plt.figure(figsize = (10,15), facecolor='white')
graphplot = 1
for column in numerical:
    if graphplot <=6:
        ax = plt.subplot(2,3,graphplot)
        sns.distplot(numerical[column])
        plt.xlabel(column, fontsize=15)
    graphplot+=1
plt.show()

Fare column has skewness and other column have classification problems

In [ ]:
# Let's check the outliers

plt.figure(figsize = (15,10), facecolor='white')
graphplot = 1
for column in numerical:
    if graphplot<=6:
        ax = plt.subplot(2,3,graphplot)
        sns.boxplot(numerical[column])
        plt.xlabel(column,fontsize=15)
    graphplot+=1
plt.show()

Outliers are present in Fare column.

In [ ]:
# removing the outliers

from scipy.stats import zscore

z = np.abs(zscore(numerical))
titanic_data = titanic_data[(z<3).all(axis=1)]
titanic_data.shape

In [ ]:
#  checking skewness again

titanic_data.skew()

Skewness is still pending in Fare column. Let's apply another method

In [ ]:
# removing the pending skewness

for index in titanic_data.skew().index:
    if titanic_data.skew().loc[index]>0.5:
        titanic_data[index] = np.sqrt(titanic_data[index])
    if titanic_data.skew().loc[index]<-0.5:
        titanic_data[index] = np.cbrt(titanic_data[index])

In [ ]:
#  checking skewness again..

titanic_data.skew()

Skewness is still pending in Fare column. Let's apply Power Transformer to remove it...

In [ ]:
# Let's apply Power Transformer to remove the skewness

from sklearn.preprocessing import PowerTransformer

features = ["Pclass","Age", "SibSp", "Parch","Fare"]

scaler = PowerTransformer(method='yeo-johnson')
titanic_data[features] = scaler.fit_transform(titanic_data[features].values)

In [ ]:
titanic_data.skew()

It seems that skewness has gone from Fare column. Now Let's deal with categorical column.

In [ ]:
# separate the input and output

x = titanic_data.drop("Survived", axis=1)
y = titanic_data["Survived"]

In [ ]:
#checking the correlation between features and target

x.corrwith(y)

In [ ]:
# Correlation analysis by ploting a graph between features and target

x.corrwith(y).plot(kind='bar',grid=True, figsize=(8,5),rot=60,title="Correlation Between Features & Targer")

2 features is correlated in negative and 3 features in correlated in positive but not any features is highly correlated to the target. Now, let's check the multicollinearity between features

In [ ]:
# Multicollinearity between features

corr = titanic_data.corr()
corr

In [ ]:
# Let's plot a heatmap for better analysis the multicollinearity

plt.figure(figsize = (10,10))
sns.heatmap(corr,cbar=True,square=True,annot=True,cbar_kws={"shrink":.82},
            annot_kws={"size":10},cmap="Blues",fmt=".2f",linewidths=.2)
plt.xlabel("Multicollinearity between features",fontsize=20)
plt.show()

Now it is clear that no any single feature is correlated to another feature. Let's check VIF Score for more analysis.

In [ ]:
# Converting categorical column to numerical column using OneHotEncoder

cat_to_conv = ["Sex", "Cabin", "Embarked"]

titanic_data = pd.get_dummies(titanic_data, columns=cat_to_conv)

titanic_data.shape

In [ ]:
# Dataset after converting categorical columns to numerical columns

titanic_data.head()

In [ ]:
# Let's separate the input and output again..

X = titanic_data.drop(columns = ["Survived"],axis=1)
Y = titanic_data["Survived"]

In [ ]:
# Let's do Scaling first before proceeding to the VIF Score

from sklearn.preprocessing import StandardScaler

scalar = StandardScaler()
x_scaled = scalar.fit_transform(X)

x_scaled

In [ ]:
# Let's check the VIF Score first..

from statsmodels.stats.outliers_influence import variance_inflation_factor

vif = pd.DataFrame()
vif["VIF Score"] = [variance_inflation_factor(x_scaled,i) for i in range(x_scaled.shape[1])]
vif["Features"] = X.columns
                    
vif.head(5)

All the VIF Scores are less than 5. Now it is cross verified that multicollinearity is not exists between the features. Let's do the analysis of target column now.

In [ ]:
# Target Column (Survived) Analysis

titanic_data["Survived"].value_counts()

In [ ]:
# Let's plot countplot for better analysis

sns.countplot(titanic_data["Survived"])

It seems that class is not balance. Let's check the percentage of each using pie chart

In [ ]:
# let's plot a pie chart to check the contribution of each

titanic_data["Survived"].value_counts().plot.pie(figsize=(5,5), autopct="%1.2f",startangle=90,
                                                 labels=["Not Survived","Survived"],fontsize=20)

There is big difference between classes. Let's do the Oversampling to balance the classes

In [ ]:
# Oversampling for class Imbalance


from imblearn.over_sampling import SMOTE

SM = SMOTE()

X,Y = SM.fit_resample(X,Y)

Y.value_counts()

In [ ]:
# Let's check countplot again...

sns.countplot(Y)

Now It is clear that class are balanced.

In [ ]:
# Let's do scaling again for the latest oversampled data

from sklearn.preprocessing import StandardScaler

sca = StandardScaler()
x_scal = sca.fit_transform(X)

x_scal

In [ ]:
# Shape of final data

titanic_data.shape

In [ ]:
# There are so many columns. So, Let's use the PCA

from sklearn.decomposition import PCA
from sklearn.metrics import explained_variance_score

x_pca = PCA()
x_pca.fit(x_scal)

var_cumu = np.cumsum(x_pca.explained_variance_ratio_)*100

k = np.argmax(var_cumu>95)
print("Number of component explaining 95% variance : ",k)

plt.xlabel("Principle Component", fontsize=15)
plt.ylabel("Cumulative Explained Variance", fontsize=15)
plt.axvline(x=k, color='k', linestyle='--')
plt.axhline(y=95, color='r', linestyle='--')
plt.plot(var_cumu)
plt.plot()

In [ ]:
# Finalysing the features with PCA for model buildings

Pca = PCA(n_components = 131)
df_x = Pca.fit_transform(x_scal)

df_x

In [ ]:
# Let's import the necessary libraries for models

from sklearn.metrics import accuracy_score, plot_roc_curve, classification_report
from sklearn.model_selection import train_test_split, cross_val_score


In [ ]:
# Let's find the best random state

def model(mod):
    max_auc = 0
    max_state = 0
    for i in range(1,201):
        x_train,x_test,y_train,y_test = train_test_split(df_x,Y,test_size = 0.25, random_state =i)
        mod.fit(x_train,y_train)
        y_pred = mod.predict(x_test)
        auc_scor = accuracy_score(y_test, y_pred)
        if auc_scor>max_auc:
            max_auc=auc_scor
            max_state=i
    print("Best Accuracy Score corresponding to ",max_state,"is ",auc_scor)
    print("Cross Validation Score : ",cross_val_score(mod,df_x,Y,cv=5).mean())
    print("Classifiation Report-->\n\n",classification_report(y_test,y_pred))

In [ ]:
# Logistic Regression

from sklearn.linear_model import LogisticRegression

LR = LogisticRegression()
model(LR)

In [ ]:
# Models Format

def models(mods):
    x_train,x_test,y_train,y_test = train_test_split(df_x,Y,test_size = 0.25, random_state =41)
    mods.fit(x_train,y_train)
    y_preds = mods.predict(x_test)
    auc_score = accuracy_score(y_test, y_preds)
    print("Best Accuracy Score corresponding to ",auc_score)
    print("Cross Validation Score : ",cross_val_score(mods,df_x,Y,cv=5).mean())
    print("Classifiation Report-->\n\n",classification_report(y_test,y_preds))          

In [ ]:
# Decision Tree Classifier

from sklearn.tree import DecisionTreeClassifier

DTR = DecisionTreeClassifier()

models(DTR)

In [ ]:
# K-Neighbors Classifier

from sklearn.neighbors import KNeighborsClassifier

KNC = KNeighborsClassifier()

models(KNC)

In [ ]:
# Gradient Boosting Classifier

from sklearn.ensemble import GradientBoostingClassifier

GBC = GradientBoostingClassifier()

models(GBC)

In [ ]:
# Bagging Classifier

from sklearn.ensemble import BaggingClassifier

BC = BaggingClassifier()

models(BC)

In [ ]:
# Random Forest Classifier

from sklearn.ensemble import RandomForestClassifier

RFC = RandomForestClassifier()

models(RFC)

****Let's check ROC AUC Curve for the fitted model****

In [ ]:
x_train,x_test,y_train,y_test = train_test_split(df_x,Y,test_size = 0.25, random_state =41)

In [ ]:
disp = plot_roc_curve(LR, x_test, y_test)
plot_roc_curve(DTR, x_test, y_test, ax=disp.ax_)
plot_roc_curve(KNC, x_test, y_test, ax=disp.ax_)
plot_roc_curve(BC, x_test, y_test, ax=disp.ax_)
plot_roc_curve(RFC, x_test, y_test, ax=disp.ax_)
plot_roc_curve(GBC, x_test, y_test, ax=disp.ax_)

plt.legend(prop={"size":10},loc="lower right")
plt.show()

**After analyzing ROC AUC Curve the Accuray Scores & Cross Validatio Scores of above 6 models, I choose Logistic Regression becasue this is the only model which is not getting the difference in accuracy score after cross validation and It is looking stable with it's score. Now let's try to increase the score by applying hyperparameter tuning**

In [ ]:
# Hyperparameter Tuning with Logistic Regression

from sklearn.model_selection import GridSearchCV

param_grid = {"penalty" : ['l1','l2'], "intercept_scaling" : [1,2,3],
             "random_state" : [81,131,151], "max_iter" : [100,200]}

grid_srch = GridSearchCV(LR, param_grid=param_grid)
grid_srch.fit(df_x,Y)
grid_srch.best_params_

**Final Model**

In [ ]:
Final_Model = LogisticRegression(intercept_scaling=1, max_iter=100, penalty='l2', random_state=81)

x_train,x_test,y_train,y_test = train_test_split(df_x,Y,test_size = 0.25, random_state =41)
Final_Model.fit(x_train,y_train)
y_preds = Final_Model.predict(x_test)
auc_scr = accuracy_score(y_test, y_preds)
print("Best Accuracy Score corresponding to ",auc_scr)
print("Classifiation Report-->\n\n",classification_report(y_test,y_preds))

**Saving The Model**

In [ ]:
import joblib

joblib.dump(Final_Model,"Final_Titanic_Project_Model.plk")